server/container_create_linux.go

```diff
@@ -1101,3 +1114,34 @@ func newLinuxContainerSecurityContext() *types.LinuxContainerSecurityContext {
 		Apparmor:         &types.SecurityProfile{},
 	}
 }
+
+func getRequestedCDIDevices(annotations map[string]string) ([]string, error) {
+	// TODO:
+	//   CDI device injection is currently requested using container annotations
+	//   with a CDI-specific key namespace. Once CRI is extended with native CDI
+	//   support, this will need to be updated accordingly.
+	_, devices, err := cdi.ParseAnnotations(annotations)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse CDI device annotations")
+	}
+
+	return devices, nil
+}
+
+func injectCDIDevices(ctx context.Context, specgen *generate.Generator, devices []string) error {
+	registry := cdi.GetRegistry()
```

Comment:
Question: Is there a way to make this registry injectable instead so that we can better test this function? I'm not sure at which level we want to expose it though.

Reply:
My plan is to add a few e2e/bats test for CDI, where we at least test the basics of the expected behavior (maybe failure for invalid annotation, failure for unresolvable devices, successful injection of a resolvable device). I think with the current test suite setup, that's the easiest way to test this and it is also more in line with the rest of the stuff here.

I did try to move the CDI-related functions to `container_create_test.go` and turn them into ginko/gomega-based tests. However, I gave up on that idea because 1) then we'd need to export those two functions to be testable (server_test vs. server package), and 2) because it looks to me like the test suite has only enough mocking for testing failing container creation, not successful ones.

Reply:
@elezar I added a few bats tests to verify the integrated CDI functionality.

Reply:
Thanks.